### PR TITLE
comp: mux: add init/free and cmd op code to mux

### PR DESCRIPTION
**Pull request by ranj063**
This will be a general purpose mux. Add the init/free
and cmd op to support passing enum values to the mux
comp.

Signed-off-by: Liam Girdwood <liam.r.girdwood@linux.intel.com>
Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

### CODE REVIEW

File: src/include/uapi/ipc/topology.h [hunk 1]
```diff
@@ -66,6 +66,12 @@ enum sof_comp_type {
 	SOF_COMP_FILEWRITE,	/**< host test based file IO */
 };
 
+enum sof_mux_flavour {
```
Comment (lbetlej) [inline src/include/uapi/ipc/topology.h]: I guess we need some SOF_COMP_DEFAULT_FLAVOR enum value that will be a default for other components.

File: src/audio/mux.c [hunk 2]
```diff
@@ -59,12 +97,94 @@ static int mux_params(struct comp_dev *dev)
...
+		return -EINVAL;
+	}
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_ENUM:
+		/*FIXME: fix logger for keywork */
+		trace_mux("mux_ctrl_set_cmd(), SOF_CTRL_CMD_ENUM, ");
+
+		/* save switch state */
+		for (j = 0; j < cdata->num_elems; j++)
```
Comment (lbetlej) [inline src/audio/mux.c]: what exactly is stored in num_elements when for cmd == SOF_CMD_ENUM ? is it a number of parameters and each chanv[j].value represents enum value of parameter[0..num_elements] ?

File: src/include/uapi/ipc/topology.h [hunk 3]
```diff
@@ -66,6 +66,12 @@ enum sof_comp_type {
...
+	SOF_COMP_MUX_GENERIC = 0,	/* generic mux */
+	SOF_COMP_MUX_CH_SEL,		/* channel selector */
+	SOF_COMP_MUX_KPB,		/* keyphrase buffer manager */
```
Comment (plbossart) [inline src/include/uapi/ipc/topology.h]: ABI level change?

File: src/audio/mux.c [hunk 4]
```diff
@@ -33,23 +33,61 @@
...
 #include <sof/audio/component.h>
 
 /* tracing */
 #define trace_mux(__e) trace_event(TRACE_CLASS_MUX, __e)
 #define trace_mux_error(__e)   trace_error(TRACE_CLASS_MUX, __e)
 #define tracev_mux(__e)        tracev_event(TRACE_CLASS_MUX, __e)
 
+struct mux_data {
+	uint32_t period_bytes;
+	uint32_t mux_value[2];
```
[the remark this answers is not in the document]
Comment (ranj063) [reply src/audio/mux.c]: @plbossart this should be changed sorry to the max channels supported.

**Comment by ranj063** [comment]
@lgirdwood @plbossart updated now. I will add the subtype bits in ww9

File: src/audio/mux.c (hunk 2 above)
Comment (lgirdwood) [reply src/audio/mux.c]: its the number of channel elements to be changed.

**Comment by lgirdwood** [comment]
@plbossart any comments on the update ?

**Comment by tlauda** [comment]
@ranj063 Is this change still valid?

**Comment by lgirdwood** [comment]
@tlauda yes, as this is muxing at the stream level rather than selecting channels. i.e. it can be used to direct output to different DAIs from a single PCM. @ranj063 CI scan failing, can you check.

**Comment by tlauda** [comment]
@lgirdwood @ranj063 It seems that @akloniex has done the full implementation in #1275. Should we close this one?

**Pull request closed without merging**